src/FSharp.Data.GraphQL.Server.Middlewares/DirectiveChooser.fs

```diff
@@ -0,0 +1,105 @@
+﻿namespace FSharp.Data.GraphQL.Server.Middlewares
+
+open FSharp.Data.GraphQL.Ast
+
+/// A function that checks if a directive should be used in the exection of a query, or changed to a new directive.
+type DirectiveChooser = Directive -> Directive option
+
+/// Basic operations on DirectiveChoosers.
+[<RequireQualifiedAccess>]
+module DirectiveChooser =
+    /// Apply a chooser to a directive.
+    let apply (directive : Directive) (chooser : DirectiveChooser) = 
+        chooser directive
+
+    /// Builds a chooser that, given a Directive x, returns Some x.
+    let keep : DirectiveChooser = 
+        let chooser = fun directive -> Some directive
+        chooser
+
+    /// Builds a chooser that, for any Directive, returns None.
+    let fallback : DirectiveChooser = 
+        let chooser = fun _ -> None
+        chooser
+
+    /// Builds a chooser that, when run, runs actual chooser, and if it returns Some directive x, maps
+    /// x directive using mapper function to y directive, and return Some y. Otherwise, returns None.
+    let map (mapper : Directive -> Directive) (actual : DirectiveChooser) : DirectiveChooser =
+        let chooser = fun directive ->
+            match actual directive with
+            | Some d -> mapper d |> keep
+            | None -> None
+        chooser
+
+    /// Builds a chooser that, given a Directive x, apply the condition filter function to x,
+    /// and if it returns true, returns Some x. Otherwise, returns None.
+    let keepWhen (condition : Directive -> bool) : DirectiveChooser = 
+        let chooser = fun directive ->
+            if condition directive
+            then keep directive
+            else fallback directive
+        chooser
+
+    /// Builds a chooser that, given a Directive x, apply the condition filter function to x,
+    /// and if it returns true, returns None. Otherwise, returns Some x.
+    let fallbackWhen (condition : Directive -> bool) : DirectiveChooser =
+        let chooser = fun directive ->
+            if condition directive
+            then fallback directive
+            else keep directive
+        chooser
+
+    /// Builds a chooser that, given a Directive x, if x.Name equals given name, returns None.
+    /// Otherwise, returns Some x.
+    let fallbackByName name = fallbackWhen (fun d -> d.Name = name)
+
+    /// Builds a chooser that, given a Directive x, if x.Name is 'defer', returns None.
+    /// Otherwise, returns Some x.
+    let fallbackDefer = fallbackByName "defer"
```

Comment:
On a second thought, I'm not sure if we need all of those helpers on the `DirectiveChooser` module. Can be useful to make readable code, but things can be much simpler if we just apply the chooser function directly.